mem/table.go

```diff
@@ -2,86 +2,284 @@ package mem
 }
 
-// TransformUp implements the Transformer interface.
-func (t *Table) TransformUp(f sql.TransformNodeFunc) (sql.Node, error) {
-	return f(t)
+var _ sql.RowIter = (*tableIter)(nil)
+
+func (i *tableIter) Next() (sql.Row, error) {
+	row, err := i.getRow()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, f := range i.filters {
+		result, err := f.Eval(sql.NewEmptyContext(), row)
+		if err != nil {
+			return nil, err
+		}
+
+		if result != true {
+			return i.Next()
+		}
+	}
+
+	return projectOnRow(i.columns, row), nil
+}
+
+func (i *tableIter) Close() error {
+	if i.indexValues == nil {
+		return nil
+	}
+
+	return i.indexValues.Close()
+}
+
+func (i *tableIter) getRow() (sql.Row, error) {
+	if i.indexValues != nil {
+		return i.getFromIndex()
+	}
+
+	if i.pos >= len(i.rows) {
+		return nil, io.EOF
+	}
+
+	row := i.rows[i.pos]
+	i.pos++
+	return row, nil
+}
+
+func projectOnRow(columns []int, row sql.Row) sql.Row {
+	if len(columns) < 1 {
+		return row
+	}
+
+	projected := make([]interface{}, len(columns))
+	for i, selected := range columns {
+		projected[i] = row[selected]
+	}
+
+	return projected
+}
+
+func (i *tableIter) getFromIndex() (sql.Row, error) {
+	data, err := i.indexValues.Next()
+	if err != nil {
+		return nil, err
+	}
+
+	value, err := decodeIndexValue(data)
+	if err != nil {
+		return nil, err
+	}
+
+	return i.rows[value.Pos], nil
+}
+
+type indexValue struct {
+	Key string
+	Pos int
+}
+
+func decodeIndexValue(data []byte) (*indexValue, error) {
+	dec := gob.NewDecoder(bytes.NewReader(data))
+	var value indexValue
+	if err := dec.Decode(&value); err != nil {
+		return nil, err
+	}
+
+	return &value, nil
 }
 
-// TransformExpressionsUp implements the Transformer interface.
-func (t *Table) TransformExpressionsUp(f sql.TransformExprFunc) (sql.Node, error) {
-	return t, nil
+func encodeIndexValue(value *indexValue) ([]byte, error) {
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	if err := enc.Encode(value); err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
 }
 
 // Insert a new row into the table.
 func (t *Table) Insert(ctx *sql.Context, row sql.Row) error {
-	if len(row) != len(t.schema) {
-		return sql.ErrUnexpectedRowLength.New(len(t.schema), len(row))
+	if err := checkRow(t.schema, row); err != nil {
+		return err
+	}
+
+	key := string(t.keys[t.insert])
+	t.insert++
+	if t.insert == len(t.keys) {
+		t.insert = 0
 	}
 
-	for idx, value := range row {
-		c := t.schema[idx]
+	t.partitions[key] = append(t.partitions[key], row)
+	return nil
+}
+
+func checkRow(schema sql.Schema, row sql.Row) error {
+	if len(row) != len(schema) {
+		return sql.ErrUnexpectedRowLength.New(len(schema), len(row))
+	}
+
+	for i, value := range row {
+		c := schema[i]
 		if !c.Check(value) {
 			return sql.ErrInvalidType.New(value)
 		}
 	}
 
-	t.data = append(t.data, row.Copy())
 	return nil
 }
 
-func (t Table) String() string {
+// String implements the sql.Table inteface.
+func (t *Table) String() string {
```

Comment:
This is still needed at this level or can be implemented at ResolvedTable level?

Reply:
it can be implemented at ResolvedTable level, but the table implementation can provide a better information since there's no way to know if there are filters, projections or indexes set in the table from the ResolvedTable